subway_prediction_ml/terminal_modes_adder.py

```diff
@@ -0,0 +1,79 @@
+import pandas as pd
+from sklearn.base import BaseEstimator, TransformerMixin
+
+from location_id_translator import LocationIdTranslator
+
+
+class TerminalModesAdder(BaseEstimator, TransformerMixin):
+    def __init__(self, locations_frame, patterns_frame, terminals_frame):
+        self.locations_frame = locations_frame
+        self.patterns_frame = patterns_frame
+        self.terminals_frame = terminals_frame
+
+    def fit(self, vehicle_datapoints, y=None):
+        return self
+
+    def transform(self, vehicle_datapoints, y=None):
+        frame_with_terminal_modes = pd.merge(
+            self._vehicle_datapoints_with_terminals(vehicle_datapoints),
+            right=self._load_terminal_datapoints(),
+            left_on=["generation", "terminal_gtfs_id"],
+            right_on=["generation", "terminal_stop_id"],
+            how="inner"
+        ).drop(
+            ["terminal_stop_id", "timestamp_y", "pattern_id"],
+            axis=1
+        ).rename({"timestamp_x": "timestamp"}, axis="columns")
+        return frame_with_terminal_modes
+
+    # Build a dataframe mapping each pattern ID to the corresponding terminal
+    # stop GTFS ID
+    def _gtfs_id_for_terminals(self):
+        patterns = self._load_patterns()
+        locations = LocationIdTranslator(self.locations_frame)
+        patterns["terminal_gtfs_id"] = patterns["terminal_stop"] \
+            .apply(lambda loc_id: locations.get(loc_id))
+        patterns = patterns.dropna().drop("terminal_stop", axis=1)
+        return patterns
+
+    # Build a dataframe mapping each pattern ID to the corresponding terminal
+    # stop location ID
+    def _load_patterns(self):
+        dropped_frame = self. \
+            patterns_frame. \
+            filter(['pattern_id', 'terminal_stop'], axis=1). \
+            dropna()
+        dropped_frame["terminal_stop"] = dropped_frame["terminal_stop"] \
+            .apply(lambda str: str.split("|")[0])
+        return dropped_frame
+
+    # Build a dataframe with all logged terminal-mode datapoints
+    def _load_terminal_datapoints(self):
+        filtered_terminals_frame = self.terminals_frame.filter([
+            'automatic',
+            'generation',
+            'terminal_stop_id',
+            'timestamp'
+        ], axis=1)
+        filtered_terminals_frame['terminal_stop_id'] = \
+            filtered_terminals_frame['terminal_stop_id'].astype('str')
+        return filtered_terminals_frame
+
+    def _translate_terminals(self, vehicle_datapoints):
+        return vehicle_datapoints.replace(
+            ['70060', '70093', '70838'],
```

Comment:
Aha, good find! I think this demonstrates the problem we're having, but I have to say I'm ambivalent about this solution.

It looks like the `ml_datapoint terminal_datapoint` is logged with `stop_id=[...]` and that stop ID is one of `"70061"`, `"70094"`, etc. I'd argue this is actually wrong, since internally the stop IDs of our terminals are represented as `:alewife`, `:ashmont`, etc. That would have been better to log, since as this issue points out, there are multiple stop IDs for each terminal. What's more, the different stop_ids _matter_. E.g., we want to predict _arrivals_ to `"70060"`, and _departures_ from `"70059"`.

The reason we're logging this terminal info is for stuff like whether the terminal is on "automatic mode" (ATD) or not, for a given vehicle datapoint (to know if its `offset_departure_seconds` is relevant). Looking at the info that we have in `vehicle_datapoints`, I think `pattern_id` most semantically captures what terminal it's going to, and hence which ATD mode applies.

Ultimately the stop_id matters here, it looks like, because initially you tried to dynamically determine the stop_id from the pattern, but as you've found that's a little inconsistent (for example, we've labeled the `:ashmont` terminal data as `"70094"` (the stop ID for boarding the inbound trip), but the final stop_id is `"70093"` (from the stop ID for unloading the outbound trip). I think had we correctly logged this as `terminal=ashmont` from the beginning, you wouldn't have been misled into trying this approach.

So back to my ambivalence: I see two ways forward.

1) This works, so let's keep it. The hacky replacement of `"70059"` / `"70060"` is basically internal to this step in the pipeline so it doesn't mess stuff up outside of it. Or:

2) Go back to first principles: squint at the `terminal_dataframe`'s `"70038"` and see it as nothing more than a fuzzy `bowdoin`. Then consider that the `vehicle_datapoints` have a `pattern_id` which most semantically determines which terminal info matters for it. Then just have a short, hardcoded mapping of the pattern_id to the terminal_id. `{'10': '70038', ...}`. I believe this simplifies the class a lot: it doesn't need to take either locations _or_ patterns, or do much of this merging.

Thoughts?

Reply:
Okay, per an offline conversation, I'm coming around on (1) since we have >100 patterns and that would be a pain to hard code. Using just the "big 8", discards about 20% of the heavy rail examples, and those are mostly yard, storage, and short-turned ones, which are arguably the most interesting.